src/Compilers/Core/Portable/IReferenceOrISignature.cs

```diff
@@ -0,0 +1,133 @@
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Used to devirtualize Dictionary/HashSet for EqualityComparer{T}.Default
+    /// </summary>
+    internal struct IReferenceOrISignatureEquivalent : IEquatable<IReferenceOrISignatureEquivalent>
+    {
+        private readonly object? _item;
+
+        public IReferenceOrISignatureEquivalent(IReference item)
+        {
+            _item = item;
+        }
+
+        public IReferenceOrISignatureEquivalent(ISignature item)
+        {
+            _item = item;
+        }
+
+        public IReferenceOrISignatureEquivalent(IMethodReference item)
+        {
+            _item = item;
+        }
+
+        public bool Equals(IReferenceOrISignatureEquivalent other)
+        {
+            object? x = _item;
+            object? y = other._item;
+            if (x is null)
+            {
+                return y is null;
+            }
+            else if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+
+            return Equals(x, y);
+        }
+
+        private new static bool Equals(object? x, object? y)
+        {
+            if (x is null)
+            {
+                return y is null;
+            }
+            else if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+            else if (x is ISymbolInternal sx && y is ISymbolInternal sy)
+            {
+                return sx.Equals(sy, TypeCompareKind.ConsiderEverything);
+            }
+            else if (x is ISymbolCompareKindComparableInternal cx && y is ISymbolCompareKindComparableInternal cy)
+            {
+                return cx.Equals(cy, TypeCompareKind.ConsiderEverything);
+            }
+            else
+            {
+                return x.Equals(y);
+            }
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is IReferenceOrISignatureEquivalent refOrSig ?
+                Equals(refOrSig) :
+                Equals(_item, obj);
+        }
+
+        public override int GetHashCode() => _item?.GetHashCode() ?? 0;
+
+        public override string ToString() => _item?.ToString() ?? "null";
+
+        internal object AsObject() => _item!;
+    }
+
+    /// <summary>
+    /// Used to devirtualize ConcurrentDictionary for EqualityComparer{T}.Default and ReferenceEquals
+    /// </summary>
+    internal struct IReferenceOrISignature : IEquatable<IReferenceOrISignature>
+    {
+        private readonly object? _item;
+
+        public static implicit operator IReferenceOrISignature(IReferenceOrISignatureEquivalent d) => new IReferenceOrISignature(d.AsObject());
+
+        private IReferenceOrISignature(object? item)
+        {
+            _item = item;
+        }
```

Comment:
I can't see where this ctor is used?

Reply:
Just above in the implicit operator :)

Reply:
@benaadams Ahh I see it now :smile:
was looking below only 😆